casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,234 @@
+        _     <- incomingQueue.enqueue1(block)
+      } yield {
+        bs.put(Seq((block.blockHash, block))) wasCalled once
+        bs.contains(*) wasNever called
+        br.ackReceived(*) wasNever called
+        dagStorageWasNotModified(bds)
+        outStream should notEmit
+      }
+  }
+
+  it should "pass to output blocks with resolved dependencies" in withEnv[Task]("root") {
+    case (incomingQueue, validatedQueue, outStream, bs, br, bds) =>
+      for {
+        // Received a parent with an empty list of justifications and its child
+        a1 <- Sync[Task].delay(makeBlock())
+        a2 = makeBlock(List(a1.blockHash))
+
+        // Put the parent and child in the input queue
+        _ <- incomingQueue.enqueue1(a2)
+        _ <- incomingQueue.enqueue1(a1)
+
+        // Dependencies of the child (its parent) have not yet been resolved,
+        // so only the parent goes to the output queue, since it has no dependencies
+        a1InOutQueue <- outStream.take(1).compile.lastOrError
+
+        // A1 is now validated (e.g. in BlockProcessor)
+        _ <- validatedQueue.enqueue1(a1)
+
+        // All dependencies of child A2 are resolved, so it also goes to the output queue
+        a2InOutQueue <- outStream.take(1).compile.lastOrError
+      } yield {
+        val bsPutCaptor = ArgCaptor[Seq[(BlockHash, BlockMessage)]]
+        bs.put(bsPutCaptor) wasCalled twice
+        bsPutCaptor.values should contain allOf (Seq((a1.blockHash, a1)), Seq((a2.blockHash, a2)))
+
+        val bsContainsCaptor = ArgCaptor[Seq[BlockHash]]
+        bs.contains(bsContainsCaptor) wasCalled 3.times
+        bsContainsCaptor.values should contain allOf (Seq(a1.blockHash), Seq(a2.blockHash))
+
+        val brAckReceivedCaptor = ArgCaptor[BlockHash]
+        br.ackReceived(brAckReceivedCaptor) wasCalled twice
+        brAckReceivedCaptor.values should contain allOf (a1.blockHash, a2.blockHash)
+
+        dagStorageWasNotModified(bds)
+        a1InOutQueue shouldBe a1.blockHash
+        a2InOutQueue shouldBe a2.blockHash
+      }
+  }
+
+  private def blockDagStorageMock[F[_]: Applicative]: BlockDagStorage[F] = {
+    val emptyDag = DagRepresentation(Set(), Map(), SortedMap(), DagMessageState(), Map())
+    mock[BlockDagStorage[F]].getRepresentation returnsF emptyDag
+  }
+
+  private def blockRetrieverMock[F[_]: Applicative]: BlockRetriever[F] =
+    mock[BlockRetriever[F]].ackReceived(*) returns ().pure[F]
+
+  private def blockStoreMock[F[_]: Sync]: BlockStore[F] = {
+    val state  = Ref.unsafe[F, Map[BlockHash, BlockMessage]](Map())
+    val bsMock = mock[BlockStore[F]]
+    bsMock.contains(*) answers { keys: Seq[BlockHash] =>
+      state.get.map(s => Seq(s.contains(keys.head)))
+    }
+    bsMock.put(*) answers { kvPairs: Seq[(BlockHash, BlockMessage)] =>
+      state.update(s => kvPairs.foldLeft(s) { case (acc, item) => acc + item })
+    }
+    bsMock
+  }
+
+  import fs2._
+
+  private def withEnv[F[_]: Concurrent: Sync: Log](shardId: String)(
```

Comment:
`Concurrent` already extends `Sync`. It can be seen as a hierarchy which gives you more and more capabilities. E.g. with Sync you can suspend computation and with Concurrent which includes suspend also enables to start multiple "threads".

Reply:
You are right. I always forgot to look at [this diagram](https://camo.githubusercontent.com/f2df6e72744e7d050c91ec30bb25341cc4a7bdc4c3ede0984be3074885de3ea3/68747470733a2f2f63646e2e7261776769742e636f6d2f74706f6c656361742f636174732d696e666f677261706869632f6d61737465722f636174732e7376673f63616368654275737465723d33) :slightly_smiling_face:

Reply:
You can also go to definition in IntelliJ and see inheritance directly in the source.